### DISCUSSION TITLE
Dataflow code

### DISCUSSION THREAD

**Discussion by kaveh808** [kaveh808/kons-9#26] [Ideas]
One very popular and common feature of 3D systems these days is a data flow model. Houdini is probably the best example.

In Lisp, this is usually represented by having a computed-slot feature in classes.

If this is something we want to have, should we roll our own or use an existing package?

I have implemented a very preliminary and not terribly elegant version of this in the file `procedural.lisp` as a proof of concept.

I'm thinking of Kenny Tilton's Cells package as one possibility. There are also others. Does anyone have any experience with any of these, or any thoughts?

As is my usual way, I prefer the simplest smallest package that will do the job.

**Comment by kaveh808** [comment]
Some options:

https://github.com/kennytilton/cells

https://gitlab.common-lisp.net/computed-class/computed-class

**Comment by loskool** [comment]
Is the idea simply to ensure that changes to some kind of "parameter object" propagate to the objects that depend on those parameters? Or is parallelization a concern too?  I'd like to contribute by reviewing some of these "reactive" or "dataflow" systems, but knowing basically nothing about conventions in 3d design systems, I could use a little input about what kinds of features I should be evaluating. 

I'm also reminded of Radul and Sussman's 2009 work on the "propagator network" model of computation. This model gives you, in addition to freedom from temporal considerations, natural parallelization as well as "data provenance". Is provenance something that people care about in this space? I can imagine it could be but really don't have the literacy here to make a judgement.

**Comment by kaveh808** [comment]
Basically data flowing along a dependency graph. The most advanced system is Houdini. Maybe find a video demonstrating their node graphs. The entire system is built on that.

**Comment by awolven** [comment]
I have Adhoc, Another Declarative Hierarchical Object-Centric CLOS
Customization.  It's a lazy-evaluated CLOS with computed-slots and
generative subobjects.

***@***.***:awolven/adhoc.git
